### ISSUE TITLE
Remaining graphical issues in Saints Row: The Third

### ISSUE THREAD

**Issue by MarioSonic2987** [RPCS3/rpcs3#11725]
**Update (20/1/2024)**: Enabling WCB and RCB fixes broken water on the distance.
**Update (24/9/2022)**: OpenGL regression is fixed.
#  Quick summary
~~Broken water reflections in Vulkan.~~ Needs Write Color Buffers and Read Color Buffers to fix this glitch.

## Details

~~If you use Vulkan, it has broken water reflections. Anisotropic filter set to 16x improves a little bit.~~

### Configuration needed
It needs Write Color Buffers enabled and Vertex Cache disabled to fix missing 3D graphics and the Map menu UI, respectively.

### Log

[Log](https://github.com/RPCS3/rpcs3/files/8405288/RPCS3.log.gz)

## Screenshots
![image](https://user-images.githubusercontent.com/20027681/161443560-1e36c20a-669a-4815-a1ae-dcd03eefb0d1.png)

## RSX captures
[Lighting Vulkan](https://drive.google.com/file/d/1MdcMYxHGpRwBVTZl3GwchW-MHvJH-cpt/view?usp=sharing)
[Water Vulkan](https://drive.google.com/file/d/1HW1DQeR7SguiFjNJW9lsgdQUX9el-zQH/view?usp=sharing)
### Specs:
* Windows 10
* Intel Core i9-9900KF
* Nvidia GeForce GTX 970 (512.15)

#### Savedata
[Savedata](https://drive.google.com/file/d/15Ggut2hHiEZNdXYbLzw8uDQLIj3nE4-w/view?usp=sharing)

**Comment by MarioSonic2987** [comment]
Saints Row: Gat out of Hell throws this fatal error when using OpenGL:
```
Unhandled GL format 0x83F1
(in file D:\a\1\s\rpcs3\Emu\RSX\GL\GLHelpers.h:1504[:26], in function texture)
```
[Log OpenGL](https://github.com/RPCS3/rpcs3/files/8476163/Log.GOOH.12-4-22.GL.zip)

**Comment by MarioSonic2987** [comment]
0.0.20-13232 ```(PR #11445)``` was the last build with correct graphics on OpenGL.
![image](https://user-images.githubusercontent.com/20027681/163724061-df1448f7-7688-4db6-b6f5-b3d6e70ca2f7.png)

0.0.20-13234 ```(PR #11443)``` there was a crash. It couldn't get ingame with OpenGL.

0.0.20-13282 ```(PR #11525)``` fixed that crash. Graphics were regressed.
![image](https://user-images.githubusercontent.com/20027681/163724243-c894f154-10f1-4b88-ad10-78ea7b839409.png)

**Comment by nitroos-ai** [comment]
Try #11833

**Comment by MarioSonic2987** [comment]
PR #11833 fixes crashing in Saints Row: Gat out of Hell. Graphics are still broken.

**Comment by MarioSonic2987** [comment]
PR #12020 makes Saints Row games crash randomly when going ingame with no fatal error.
[Log Gat out of Hell Broken](https://github.com/RPCS3/rpcs3/files/8790892/RPCS3.log.gz)

**Comment by MarioSonic2987** [comment]
In 0.0.23-13883, crashes are no longer present.

**Comment by MarioSonic2987** [comment]
Broken graphics in OpenGL are fixed in 0.0.24-14175.
![image](https://user-images.githubusercontent.com/20027681/191744820-a1f6ec9a-4f6b-46fa-91f2-5d81a84bbd6e.png)
![image](https://user-images.githubusercontent.com/20027681/191745359-89a30487-672d-46b8-88fa-4ee46cbc5b07.png)

**Comment by MarioSonic2987** [comment]
As of 0.0.28-15212, distant water reflection is now fixed, although it's now flickering in Vulkan. However, water is now waving weirdly when you're close to it.

**Vulkan**:
https://github.com/RPCS3/rpcs3/assets/20027681/a49d2302-321d-4d92-86a1-e3e87cdd02b0

**OpenGL**:
https://github.com/RPCS3/rpcs3/assets/20027681/21a6c040-de5f-4a18-aaf8-5a2c3d2a64da

**Comment by MarioSonic2987** [comment]
Some elements are drawn through the objects, like the lights on the plane in the beginning of the game. Same with Saints Row IV and Gat out of Hell.

**RPCS3**:
![BugSRTT](https://github.com/RPCS3/rpcs3/assets/20027681/f74d95db-8348-4ed6-bf15-ff97feaace95)

**PS3**:
![image](https://github.com/RPCS3/rpcs3/assets/20027681/ed40bdfa-1512-4d4b-95af-5475a4f6c3a2)

They are also present on vehicle headlights.
![image](https://github.com/RPCS3/rpcs3/assets/20027681/e1d3c3d2-cba5-49b2-bc00-68e34030aecc)

**Comment by MarioSonic2987** [comment]
Enabling Write Color Buffers and Read Color Buffer fixes water on the distance.
![image](https://github.com/RPCS3/rpcs3/assets/20027681/05aae885-39aa-4397-a03c-0c8425f11ae4)

Write Color Buffers only:
![image](https://github.com/RPCS3/rpcs3/assets/20027681/f836e43f-9227-47f8-9be9-c6ab33a3438a)


The shaking water bug is still present:
https://github.com/RPCS3/rpcs3/assets/20027681/ae644525-7eda-4343-beed-b77c569ff7f5

**Comment by kd-11** [comment]
> Enabling Write Color Buffers and Read Color Buffer fixes water on the distance. ![image](https://private-user-images.githubusercontent.com/20027681/298286562-05aae885-39aa-4397-a03c-0c8425f11ae4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU3OTAxMDAsIm5iZiI6MTcwNTc4OTgwMCwicGF0aCI6Ii8yMDAyNzY4MS8yOTgyODY1NjItMDVhYWU4ODUtMzlhYS00Mzk3LWEwM2MtMGM4NDI1ZjExYWU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIwVDIyMzAwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiOGM5MDNhNmE5OTY4ZmUxMjcyMTFhNjdiZTcxMWY2MDViYTE2YzlhY2E1ODEzOTQ3YmQ3MGRhMzU4OGY1YTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QzehvIs3-BEB1unEta7EU88H0fJjtyzeHa-R5t0fsI8)
> 
> Write Color Buffers only: ![image](https://private-user-images.githubusercontent.com/20027681/298286661-f836e43f-9227-47f8-9be9-c6ab33a3438a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU3OTAxMDAsIm5iZiI6MTcwNTc4OTgwMCwicGF0aCI6Ii8yMDAyNzY4MS8yOTgyODY2NjEtZjgzNmU0M2YtOTIyNy00N2Y4LTliZTktYzZhYjMzYTM0MzhhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIwVDIyMzAwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5Njg3ZTJmNzJiYzg2Njg3ZWE3MjBhMjU5OTNmMDYyZWExNmQyODEyYWE5NzM3MDIyNTQ1MWQxMDI4ZmZjY2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MC5MwauGExoaX8FefxVhLM-LC6iLb9zEDaOt4OqThtA)
> 
> The shaking water bug is still present:
> 
>  BLES01747_recording_2024_01_20_15_05_16.mp4

First thing to check there would be SPU options. Whenever I see weird calculations breaking animation it's almost always some SPU magic.

**Comment by MarioSonic2987** [comment]
When using SPU ASMJIT, water flickers A LOT more than with SPU LLVM (Approximate XFloat). The same happens with SPU LLVM (Accurate XFloat).

https://github.com/RPCS3/rpcs3/assets/20027681/09b385f2-9ae8-4d85-9fea-f20d086a27b1

SPU LLVM (Approximate XFloat):

https://github.com/RPCS3/rpcs3/assets/20027681/b82ebfca-e32d-4948-b7b9-17d2c37ff215

**Comment by MarioSonic2987** [comment]
When you press Alt+Enter to toggle fullscreen many times, the water breaks even more.

No Alt+Enter (water is still shaking):
![image](https://github.com/user-attachments/assets/a4993e8f-1116-484b-b991-b4cbb245856f)

Pressing Alt+Enter +3 times (water has vertex explosions):
![image](https://github.com/user-attachments/assets/77aec8c6-9a57-4329-a9a3-e98a16451b0b)

**Comment by digant73** [comment]
if the just merged PR #16352 is fixing all the main issues reported in this bug report it could be better to open a new bug report reporting the remaining headlight issue. Headlight issue seems to be also present on other games (Resistance 3, Killzone 3 etc.). So other people could provide their feedbacks on that specific issue more than on a specific game

**Comment by MarioSonic2987** [comment]
Yeah. I'll close this issue and open a new one reporting the headlights glitch.
